core/matrix/batch_diagonal.cpp

```diff
@@ -0,0 +1,255 @@
+void BatchDiagonal<ValueType>::move_to(
+    BatchDiagonal<next_precision<ValueType>>* const result)
+{
+    this->convert_to(result);
+}
+
+
+namespace {
+
+
+template <typename MatrixType, typename MatrixData>
+inline void read_impl(MatrixType* mtx, const std::vector<MatrixData>& data)
+{
+    using value_type = typename MatrixType::value_type;
+    auto batch_sizes = std::vector<dim<2>>(data.size());
+    size_type ind = 0;
+    for (const auto& b : data) {
+        batch_sizes[ind] = b.size;
+        ++ind;
+    }
+    auto tmp = MatrixType::create(mtx->get_executor()->get_master(),
+                                  batch_dim<2>(batch_sizes));
+    mtx->get_executor()->run(batch_diagonal::make_fill(
+        tmp->get_values(), tmp->get_num_stored_elements(), zero<value_type>()));
+    for (size_type b = 0; b < data.size(); ++b) {
+        for (auto nnz : data[b].nonzeros) {
+            if (nnz.row == nnz.column) {
+                tmp->at(b, nnz.row) = nnz.value;
+            }
+        }
+    }
+    tmp->move_to(mtx);
+}
+
+
+}  // namespace
+
+
+template <typename ValueType>
+void BatchDiagonal<ValueType>::read(const std::vector<mat_data>& data)
+{
+    read_impl(this, data);
+}
+
+
+template <typename ValueType>
+void BatchDiagonal<ValueType>::read(const std::vector<mat_data32>& data)
+{
+    read_impl(this, data);
+}
+
+
+namespace {
+
+
+template <typename MatrixType, typename MatrixData>
+inline void write_impl(const MatrixType* mtx, std::vector<MatrixData>& data)
+{
+    std::unique_ptr<const BatchLinOp> op{};
+    const MatrixType* tmp{};
+    if (mtx->get_executor()->get_master() != mtx->get_executor()) {
+        op = mtx->clone(mtx->get_executor()->get_master());
+        tmp = static_cast<const MatrixType*>(op.get());
+    } else {
+        tmp = mtx;
+    }
+
+    data = std::vector<MatrixData>(mtx->get_num_batch_entries());
+    for (size_type b = 0; b < mtx->get_num_batch_entries(); ++b) {
+        data[b] = {mtx->get_size().at(b), {}};
+        const auto nstored = std::min(data[b].size[0], data[b].size[1]);
+        for (size_type row = 0; row < nstored; ++row) {
+            if (tmp->at(b, row) != zero<typename MatrixType::value_type>()) {
+                data[b].nonzeros.emplace_back(row, row, tmp->at(b, row));
+            }
+        }
+    }
+}
+
+
+}  // namespace
+
+
+template <typename ValueType>
+void BatchDiagonal<ValueType>::write(std::vector<mat_data>& data) const
+{
+    write_impl(this, data);
+}
+
+
+template <typename ValueType>
+void BatchDiagonal<ValueType>::write(std::vector<mat_data32>& data) const
+{
+    write_impl(this, data);
+}
+
+
+template <typename ValueType>
+std::unique_ptr<BatchLinOp> BatchDiagonal<ValueType>::transpose() const
+{
+    auto exec = this->get_executor();
+    auto trans_cpy =
+        BatchDiagonal::create(exec, gko::transpose(this->get_size()));
+
+    exec->copy(this->get_num_stored_elements(), this->get_const_values(),
+               trans_cpy->get_values());
```

Comment:
How? The size needs to be transposed.

Reply:
Isn't diagonal always square ?

Reply:
No, I thought we can allow non-square diagonal matrices, since it's not too much more work right now to enable it. And mathematically they are sometimes useful.